gtk4: fails to compile for 32-bit platforms (i686-linux)

### Nixpkgs version

- Unstable (26.05)

### Describe the bug

Issue researched with ai, and seems correct to me:
When cross-compiling gtk4 from x86_64-linux to i686-linux, the build fails
with a type mismatch in the Vulkan rendering code. On 32-bit platforms, the
C `NULL` pointer is too small to fit into the `VkImage` type (which is always
a 64-bit integer), so the compiler rejects it.
This only happens when cross-compiling to 32-bit targets — native x86_64
builds work fine because `NULL` is also 64-bit there.
The problem is in the file gsk/gpu/gskvulkanimage.c at line 1278, where it
tries to assign `NULL` to a `VkImage` field. The correct fix is to use
`VK_NULL_HANDLE` instead (which is always 64-bits wide, no matter the platform).

### Steps to reproduce

1. update flake
2. rebuild system

### Expected behaviour

should work and not fail

### Screenshots

_No response_

### Relevant log output

```console
im getting the following error while nixos rebuild:

gtk4-i686-linux> [563/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpunodeprocessor.c.o
gtk4-i686-linux> ninja: build stopped: subcommand failed.
error: builder for '/nix/store/21fj9i12spxpyskvxd3vha7arisgbn14-gtk4-4.22.4.drv' failed with exit code 1;
       last 25 log lines:
       > [546/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpuutils.c.o
       > [547/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpurenderer.c.o
       > [548/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpuscissorop.c.o
       > [549/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpuprint.c.o
       > [550/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpurenderpassop.c.o
       > [551/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpushaderop.c.o
       > [552/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpuuploadop.c.o
       > [553/2057] Compiling C object gsk/libgsk.a.p/gpu_gskvulkanbuffer.c.o
       > [554/2057] Compiling C object gsk/libgsk.a.p/gpu_gskvulkanframe.c.o
       > [555/2057] Compiling C object gtk/libgtk.a.p/meson-generated_.._text-input-unstable-v3-protocol.c.o
       > [556/2057] Compiling C object gsk/libgsk.a.p/gpu_gskvulkanimage.c.o
       > FAILED: [code=1] gsk/libgsk.a.p/gpu_gskvulkanimage.c.o
       > gcc -Igsk/libgsk.a.p -Igsk -I../gsk -I. -I.. -Igsk/gpu -I../gsk/gpu -Igdk/x11 -I../gdk/x11 -Igdk -I../gdk -Igsk/gpu/shaders -Igdk/version -Igtk/css -I/nix/store/fiiyg5xkv6rqwmrbz1lw4634kvbbphfy-glib-2.88.1-dev/include -I/nix/store/fiiyg5xkv6rqwmrbz1lw4634kvbbphfy-glib-2.88.1-dev/include/glib-2.0 -I/nix/store/h8w127ms62k84b7gqp370y8f41942ryf-glib-2.88.1/lib/glib-2.0/include -I/nix/store/npbzh4gj0kawmlnsc393xc5rv5agaqmp-graphene-1.10.8-dev/include/graphene-1.0 -I/nix/store/qk187vvmnrwmyhv9pa4xdrqkkq5vvcf2-graphene-1.10.8/lib/graphene-1.0/include -I/nix/store/93mmbbxvis1vk7ih62lrdz3bi8sl45k6-harfbuzz-13.2.1-dev/include/harfbuzz -I/nix/store/aw95pwykc5abcnk1a22i6hjd50z50vjs-pango-1.57.1-dev/include/pango-1.0 -I/nix/store/yf4ivn4498xblh3qgk9sb12gng58f59y-cairo-1.18.4-dev/include/cairo -I/nix/store/gzysc4wxzlcgi104bvqivm1n1428p994-freetype-2.14.2-dev/include/freetype2 -I/nix/store/gzysc4wxzlcgi104bvqivm1n1428p994-freetype-2.14.2-dev/include -I/nix/store/szkx93107b3i912dahac1phrpk88s41f-gdk-pixbuf-2.44.6-dev/include/gdk-pixbuf-2.0 -I/nix/store/kali1qjjgr0ysbra7ndv2vy2ibl8xdzm-fribidi-1.0.16-dev/include/fribidi -I/nix/store/fh890mswvqj8qaqvzg6q7x25aclx9vwb-libepoxy-1.5.10-dev/include -I/nix/store/kndkzai26q3zvyawg970yd5iymjmp4j1-fontconfig-2.17.1-dev/include -I/nix/store/fiiyg5xkv6rqwmrbz1lw4634kvbbphfy-glib-2.88.1-dev/include/gio-unix-2.0 -I/nix/store/ivmr9icnx3ryrq8gydh2gpfb8s63fq17-vulkan-headers-1.4.341.0/include -I/nix/store/9cqhyxly0mcxp3k9yx86fm4akyb8bqd6-libdrm-2.4.133-dev/include -I/nix/store/9cqhyxly0mcxp3k9yx86fm4akyb8bqd6-libdrm-2.4.133-dev/include/libdrm -I/nix/store/5358z11z0my2134l5nqg3a9ma7bm1mig-libpng-apng-1.6.56-dev/include/libpng16 -I/nix/store/cb57lri45y4wjzdi6ikz8lxiiddzy0hj-libtiff-4.7.1-dev/include -I/nix/store/x0dl228354mcwzz7x6nqaxdxbcf9dg0x-libjpeg-turbo-3.1.4-dev/include -I/nix/store/vd173g4myq5r3zrx99lcdd3npvgn796b-libxkbcommon-1.13.1-dev/include -I/nix/store/fw901d9dm1j4arawyx2xc7kblwvspxfs-wayland-1.25.0-dev/include -I/nix/store/9m5xl3p3hffpvv6g66cqc99kfb0lai26-libx11-1.8.13-dev/include -I/nix/store/han1fsp7gahrnnkjh8hmg5yf6pvjl41l-xorgproto-2025.1/include -I/nix/store/mz3gnglmn613hnfz4gg9m6gapsl3azpy-libxrender-0.9.12-dev/include -I/nix/store/z6d3bnyn8yv7z606467pv1zrcrh7sqz7-libxi-1.8.2-dev/include -I/nix/store/2jgms2j0mwga8mqjvfsglm3p249lqsp8-libxext-1.3.7-dev/include -I/nix/store/wz6v57nz633zv5w95v2k50p09kil36ad-libxcursor-1.2.3-dev/include -I/nix/store/zp4yphqa052v2j1pq4r40pb1bjwgz23z-libxdamage-1.1.7-dev/include -I/nix/store/0m41y3wyffaq7jwwkhw5x1rylwv6xqwp-libxfixes-6.0.2-dev/include -I/nix/store/v0vjncf4plgcfnvn6lhx5v358rayq48q-libxrandr-1.5.5-dev/include -I/nix/store/a5r07949nb7i2flp8bkf36405wm066az-libxinerama-1.1.6-dev/include -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=gnu11 -DG_LOG_USE_STRUCTURED=1 -DGLIB_VERSION_MIN_REQUIRED=GLIB_VERSION_2_84 -DGLIB_VERSION_MAX_ALLOWED=GLIB_VERSION_2_84 '-DPANGO_VERSION_MIN_REQUIRED=G_ENCODE_VERSION(1,56)' '-DPANGO_VERSION_MAX_ALLOWED=G_ENCODE_VERSION(1,56)' '-DGTK_VERSION="4.22.4"' -D_GNU_SOURCE -fPIC -DGTK_COMPILATION '-DG_LOG_DOMAIN="Gsk"' -DG_LOG_STRUCTURED=1 -fno-strict-aliasing -Wno-missing-include-dirs -Wduplicated-branches -Wduplicated-cond -Wformat=2 -Wformat-nonliteral -Wformat-security -Wignored-qualifiers -Wimplicit-function-declaration -Wlogical-op -Wmisleading-indentation -Wmissing-format-attribute -Wmissing-include-dirs -Wmissing-noreturn -Wnested-externs -Wold-style-definition -Wpointer-arith -Wshadow -Wstrict-prototypes -Wswitch-default -Wswitch-enum -Wundef -Wuninitialized -Wunused -Waddress -Warray-bounds -Wempty-body -Wenum-int-mismatch -Wimplicit -Wimplicit-fallthrough -Wimplicit-fallthrough=5 -Winit-self -Wint-to-pointer-cast -Wmain -Wmissing-braces -Wmissing-declarations -Wmissing-prototypes -Wnonnull -Woverride-init -Wpointer-to-int-cast -Wredundant-decls -Wreturn-type -Wsequence-point -Wtrigraphs -Wvla -Wwrite-strings -Wcast-align -Wnull-dereference -fvisibility=hidden -MD -MQ gsk/libgsk.a.p/gpu_gskvulkanimage.c.o -MF gsk/libgsk.a.p/gpu_gskvulkanimage.c.o.d -o gsk/libgsk.a.p/gpu_gskvulkanimage.c.o -c ../gsk/gpu/gskvulkanimage.c
       > ../gsk/gpu/gskvulkanimage.c: In function 'gsk_vulkan_image_new_for_dmabuf':
       > ../gsk/gpu/gskvulkanimage.c:1278:22: error: assignment to 'VkImage' {aka 'long long unsigned int'} from 'void *' makes integer from pointer without a cast [-Wint-conversion]
       >  1278 |       self->vk_image = NULL;
       >       |                      ^
       > [557/2057] Compiling C object gsk/libgsk.a.p/gpu_gskvulkandevice.c.o
       > [558/2057] Compiling C object gsk/libgsk.a.p/gpu_gskvulkandebugframe.c.o
       > [559/2057] Compiling C object gsk/libgsk.a.p/gpu_gskvulkanmemory.c.o
       > [560/2057] Compiling C object gtk/libgtk.a.p/meson-generated_.._a11y_atspi_atspi-action.c.o
       > [561/2057] Compiling C object gsk/libgsk.a.p/gpu_gskvulkanycbcr.c.o
       > [562/2057] Compiling C object gtk/libgtk.a.p/meson-generated_.._a11y_atspi_atspi-accessible.c.o
       > [563/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpunodeprocessor.c.o
       > ninja: build stopped: subcommand failed.
       For full logs, run:
        nix log /nix/store/21fj9i12spxpyskvxd3vha7arisgbn14-gtk4-4.22.4.drv


nix log /nix/store/21fj9i12spxpyskvxd3vha7arisgbn14-gtk4-4.22.4.drv

[537/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpucachedfill.c.o
[538/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpuclip.c.o
[539/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpudevice.c.o
[540/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpudownloadop.c.o
[541/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpuimage.c.o
[542/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpuframe.c.o
[543/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpuglobalsop.c.o
[544/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpuop.c.o
[545/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpumipmapop.c.o
[546/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpuutils.c.o
[547/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpurenderer.c.o
[548/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpuscissorop.c.o
[549/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpuprint.c.o
[550/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpurenderpassop.c.o
[551/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpushaderop.c.o
[552/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpuuploadop.c.o
[553/2057] Compiling C object gsk/libgsk.a.p/gpu_gskvulkanbuffer.c.o
[554/2057] Compiling C object gsk/libgsk.a.p/gpu_gskvulkanframe.c.o
[555/2057] Compiling C object gtk/libgtk.a.p/meson-generated_.._text-input-unstable-v3-protocol.c.o
[556/2057] Compiling C object gsk/libgsk.a.p/gpu_gskvulkanimage.c.o
FAILED: [code=1] gsk/libgsk.a.p/gpu_gskvulkanimage.c.o 
gcc -Igsk/libgsk.a.p -Igsk -I../gsk -I. -I.. -Igsk/gpu -I../gsk/gpu -Igdk/x11 -I../gdk/x11 -Igdk -I../gdk -Igsk/gpu/shaders -Igdk/version -Igtk/css -I/nix/store/fiiyg5xkv6rqwmrbz1lw4634kvbbphfy-glib-2.88.1-dev/include -I/nix/store/fiiy>
../gsk/gpu/gskvulkanimage.c: In function 'gsk_vulkan_image_new_for_dmabuf':
../gsk/gpu/gskvulkanimage.c:1278:22: error: assignment to 'VkImage' {aka 'long long unsigned int'} from 'void *' makes integer from pointer without a cast [-Wint-conversion]
 1278 |       self->vk_image = NULL;
      |                      ^
[557/2057] Compiling C object gsk/libgsk.a.p/gpu_gskvulkandevice.c.o
[558/2057] Compiling C object gsk/libgsk.a.p/gpu_gskvulkandebugframe.c.o
[559/2057] Compiling C object gsk/libgsk.a.p/gpu_gskvulkanmemory.c.o
[560/2057] Compiling C object gtk/libgtk.a.p/meson-generated_.._a11y_atspi_atspi-action.c.o
[561/2057] Compiling C object gsk/libgsk.a.p/gpu_gskvulkanycbcr.c.o
[562/2057] Compiling C object gtk/libgtk.a.p/meson-generated_.._a11y_atspi_atspi-accessible.c.o
[563/2057] Compiling C object gsk/libgsk.a.p/gpu_gskgpunodeprocessor.c.o
ninja: build stopped: subcommand failed.
```

### Additional context

my research says it should be fixed with the upstream commit:
https://gitlab.gnome.org/GNOME/gtk/-/commit/d2fa538eba97b66d2bd5ac7d27b98925e4420e2f

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 7.0.2-xanmod1, NixOS, 26.05 (Yarara), 26.05.20260515.d233902`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Lix, like Nix) 2.94.2
System type: x86_64-linux
Additional system types: i686-linux, x86_64-v1-linux, x86_64-v2-linux, x86_64-v3-linux, x86_64-v4-linux
Features: gc, signed-caches
System configuration file: /etc/nix/nix.conf
User configuration files: /home/xyz00777/.config/nix/nix.conf:/nix/store/3wc6fkw8wmmq7ph37wsgl9pxisdbfjwp-plasma-workspace-6.6.4/etc/xdg/nix/nix.conf:/nix/store/0hhrsvzf27rldm36sks04x3q0nh4np94-kglobalacceld-6.6.4/etc/xdg/nix/nix.conf:/nix/store/a388zn50fnrk7jd44xxm7dpmr5nh64qz-baloo-6.25.0/etc/xdg/nix/nix.conf:/home/xyz00777/.config/kdedefaults/nix/nix.conf:/home/xyz00777/.config/kdedefaults/nix/nix.conf:/etc/xdg/nix/nix.conf:/home/xyz00777/.local/share/flatpak/exports/etc/xdg/nix/nix.conf:/var/lib/flatpak/exports/etc/xdg/nix/nix.conf:/home/xyz00777/.nix-profile/etc/xdg/nix/nix.conf:/nix/profile/etc/xdg/nix/nix.conf:/home/xyz00777/.local/state/nix/profile/etc/xdg/nix/nix.conf:/etc/profiles/per-user/xyz00777/etc/xdg/nix/nix.conf:/nix/var/nix/profiles/default/etc/xdg/nix/nix.conf:/run/current-system/sw/etc/xdg/nix/nix.conf
Store directory: /nix/store
State directory: /nix/var/nix
Data directory: /nix/store/dkg15fqgldf1fl0asczs6fkh6yp72fxl-lix-2.94.2/share`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/store/77dbgds155bbz3vd3qywq1sii07i5ljs-source`


### Notify maintainers

@s1341 @vcunat 
---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+label%3A%220.kind%3A+bug%22+-label%3A%226.topic%3A+darwin%22+-label%3A%226.topic%3A+nixos%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.
- [x] I assert that I have read the [automation/AI policy](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#automationai-policy) and that this issue report complies with it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/gt/gtk4/package.nix
+++ b/pkgs/by-name/gt/gtk4/package.nix
@@ -95,6 +95,12 @@ stdenv.mkDerivation (finalAttrs: {
     hash = "sha256-Ub2fYMfSOmZaVWxzZMIfsuTiglZrPn4JJFXo+RAzCJM=";
   };
 
+  patches = lib.optional stdenv.hostPlatform.is32bit (fetchpatch {
+    name = "fix-32bit-VkImage-null.patch";
+    url = "https://gitlab.gnome.org/GNOME/gtk/-/commit/10d43de8f4f942cb591ada3103474bd7213425f1.patch";
+    hash = "sha256-DJIL6M3XcsjBoMO77OxNi84d1DxAphAfot3N7Nq1QqQ=";
+  });
+
   depsBuildBuild = [
     pkg-config
   ];

